tempesta_fw/sock.c

```diff
@@ -441,106 +442,111 @@ ss_tcp_process_data(struct sock *sk)
 
-			r = SS_CALL(connection_recv, conn, skb, off);
+		r = SS_CALL(connection_recv, conn, skb, off);
 
-			bh_lock_sock_nested(sk);
+		bh_lock_sock_nested(sk);
 
-			read_unlock(&sk->sk_callback_lock);
+		read_unlock(&sk->sk_callback_lock);
 
+		/*
+		 * Unlocking of the socket above creates a chance that
+		 * the socket might be closed in a parallel thread.
+		 * Now that the socket is locked again, check to see
+		 * if that has really happened. If it has, then simply
+		 * return to the caller. The party that actually closes
+		 * the socket takes care of any additional actions like
+		 * failover, etc.
+		 *
+		 * Note that the socket had not been destroyed in case
+		 * it's been closed. An extra socket reference is taken
+		 * before this function is called which prevents that.
+		 * See tcp_v4_rcv() and __inet_lookup_established().
+		 */
+		if (unlikely(sk->sk_state != TCP_ESTABLISHED))
+			return;
+
+		if (r < 0) {
+			SS_WARN("can't process app data on socket %p\n", sk);
 			/*
-			 * Unlocking of the socket above creates a chance that
-			 * the socket might be closed in a parallel thread.
-			 * Now that the socket is locked again, check to see
-			 * if that has really happened. If it has, then simply
-			 * return to the caller. The party that actually closes
-			 * the socket takes care of any additional actions like
-			 * failover, etc.
+			 * Drop connection on internal errors as well as
+			 * on banned packets.
 			 *
-			 * Note that the socket had not been destroyed in case
-			 * it's been closed. An extra socket reference is taken
-			 * before this function is called which prevents that.
-			 * See tcp_v4_rcv() and __inet_lookup_established().
+			 * ss_droplink() is responsible for calling
+			 * application layer connection closing callback
+			 * which will free all the passed and linked
+			 * with currently processed message skbs.
 			 */
-			if (unlikely(sk->sk_state != TCP_ESTABLISHED))
-				return;
-
-			if (r < 0) {
-				SS_WARN("can't process app data on socket %p\n",
-					sk);
-				/*
-				 * Drop connection on internal errors as well as
-				 * on banned packets.
-				 *
-				 * ss_droplink() is responsible for calling
-				 * application layer connection closing callback
-				 * which will free all the passed and linked
-				 * with currently processed message skbs.
-				 */
-				goto out; /* connection dropped */
-			}
-			tp->copied_seq += count;
-			processed += count;
-		}
-		else if (tcp_hdr(skb)->fin) {
-			SS_DBG("received FIN, do active close\n");
-			++tp->copied_seq;
-			__kfree_skb(skb);
-			goto out;
-		}
-		else {
-			SS_WARN("recvmsg bug: overlapping TCP segment at %X"
-				" seq %X rcvnxt %X len %x\n",
-				tp->copied_seq, TCP_SKB_CB(skb)->seq,
-				tp->rcv_nxt, skb->len);
-			__kfree_skb(skb);
+			goto out; /* connection dropped */
 		}
+		tp->copied_seq += count;
+		processed += count;
+
+		if (tcp_fin)
+			goto data_fin;
+		continue;
+lone_fin:
+		__kfree_skb(skb);
+data_fin:
```

Comment:
The function control flow becomes to complex - too many gotos, labels and booleans. Sometimes it's better to use if statements and do little copy-paste (e.g. there is not a big deal to copy-paste ```SS_DBG(".."); ++tp->copied_seq;``` for FIN processing.